### ISSUE TITLE
`exercism upgrade` fails with `text file busy`

### ISSUE THREAD

**Issue by jbaiter** [exercism/cli#288]
I just tried to upgrade the tool from the command-line and got the following error:

```
$ exercism upgrade
Downloading exercism-linux-64bit.tgz
2016/02/01 09:30:13 open /home/jbaiter/.bin/exercism: text file busy
```

**Comment by kytrinyx** [comment]
I'm not sure what's happening with this. @Tonkpils @lcowell can either of you reproduce?

**Comment by jbaiter** [comment]
I just tried it on my Laptop (using the same system, Debian unstable) and could reproduce it. I uploaded a log of all syscalls (via `strace`) here: https://gist.github.com/jbaiter/8f473361245ad3649c78

**Comment by lcowell** [comment]
I was able to upgrade my client without difficulty on OS X. @jbaiter What kind of filesystem is `/home/jbaiter` ?

I bet this is a linux specific issue. We might need to update the upgrade process to:
1) download the new file
2) rename the current binary to old
3) rename the new binary to current
4) remove the old binary

**Comment by kytrinyx** [comment]
Yeah, it would be worth creating a test-build on an older version with that process (so that it doesn't just say "You're up to date" when you install the test build).

**Comment by Tonkpils** [comment]
> 1) download the new file
> 2) rename the current binary to old
> 3) rename the new binary to current
> 4) remove the old binary

Yeah, so I'm able to confirm this. The one issue we will run into is with Windows as well during step 4. Windows won't allow removal of the old binary since it's technically still in use. One way we can go about this is actually hide the old binary and remove it attempt removal at the beginning of every upgrade of the old renamed binary. 

I'll give this a try

**Comment by jbaiter** [comment]
I just toyed around with the `installTgz` function in `release_utils.go` and I think I managed to fix the issue, at least on my machine. All I had to do was remove any existing files that would be overwritten.

Caveat: I don't know any Go, so this might be the wrong way to check for file existence/removing them:

```go
func installTgz(source *bytes.Reader, dest string) error {
        gr, err := gzip.NewReader(source)
        if err != nil {
                return err
        }
        defer gr.Close()

        tr := tar.NewReader(gr)
        for {
                hdr, err := tr.Next()
                if err == io.EOF {
                        break
                }
                if err != nil {
                        return err
                }

                // FIXED: Remove any existing files before attempting to overwrite them
                if _, err := os.Stat(dest); err == nil {
                    os.Remove(dest)
                }

                fileCopy, err := os.OpenFile(dest, installFlag, hdr.FileInfo().Mode())
                if err != nil {
                        return err
                }
                defer fileCopy.Close()

                if _, err = io.Copy(fileCopy, tr); err != nil {
                        return err
                }
        }

        return nil
}
```

**Comment by Tonkpils** [comment]
> All I had to do was remove any existing files that would be overwritten.

That would assume that the write for the `upgrade` will succeed. If we remove the binary (which won't work on Windows) and the writing of the file fails, we will be in an inconsistent state. There's a couple of edge cases for this. 

That's a good start though :+1: If you'd like to take a crack at making a PR for this I'd definitely be open to pairing with you.

**Comment by jbaiter** [comment]
> That would assume that the write for the upgrade will succeed. If we remove the binary (which won't work on Windows) and the writing of the file fails, we will be in an inconsistent state. There's a couple of edge cases for this. 

My changes don't apply to Windows anyway, since it uses the `installZip` function, which remains unchanged. But you're completely right about the inconsistent state in case the writing fails. I just pushed another commit that works around this case by creating a backup copy and recovering from that in case of an error.

I just opened a PR and am open for further changes/edge cases :-)
